diff --git a/frameworks/base/media/jni/Android.mk b/frameworks/base/media/jni/Android.mk
index 659a315..fc4c0a0 100644
--- a/frameworks/base/media/jni/Android.mk
+++ b/frameworks/base/media/jni/Android.mk
@@ -11,7 +11,6 @@ LOCAL_SRC_FILES:= \
     android_media_MediaHTTPConnection.cpp \
     android_media_MediaMuxer.cpp \
     android_media_MediaPlayer.cpp \
-    android_media_ExtMediaPlayer.cpp \
     android_media_MediaRecorder.cpp \
     android_media_MediaScanner.cpp \
     android_media_MediaMetadataRetriever.cpp \
@@ -23,6 +22,11 @@ LOCAL_SRC_FILES:= \
     android_mtp_MtpDevice.cpp \
     android_mtp_MtpServer.cpp \
 
+ifeq ($(BOARD_USES_QCOM_HARDWARE),true)
+LOCAL_SRC_FILES += \
+    android_media_ExtMediaPlayer.cpp
+endif
+
 LOCAL_SHARED_LIBRARIES := \
     libandroid_runtime \
     libnativehelper \
diff --git a/frameworks/base/media/jni/android_media_MediaPlayer.cpp b/frameworks/base/media/jni/android_media_MediaPlayer.cpp
index a6a9f35..0175951 100644
--- a/frameworks/base/media/jni/android_media_MediaPlayer.cpp
+++ b/frameworks/base/media/jni/android_media_MediaPlayer.cpp
@@ -37,7 +37,9 @@
 #include "utils/KeyedVector.h"
 #include "utils/String8.h"
 #include "android_media_Utils.h"
+#ifdef QCOM_HARDWARE
 #include "android_media_ExtMediaPlayer.h"
+#endif
 
 #include "android_os_Parcel.h"
 #include "android_util_Binder.h"
@@ -696,6 +698,7 @@ android_media_MediaPlayer_native_setup(JNIEnv *env, jobject thiz, jobject weak_t
     // create new listener and give it to MediaPlayer
     sp<JNIMediaPlayerListener> listener = new JNIMediaPlayerListener(env, thiz, weak_this);
     mp->setListener(listener);
+#ifdef QCOM_HARDWARE
     if (JNIExtMediaPlayerListener::checkExtMedia(env, thiz)) {
       ALOGD("QCMediaPlayer mediaplayer present");
        sp<JNIExtMediaPlayerListener> extmedialistener = new JNIExtMediaPlayerListener(
@@ -704,6 +707,7 @@ android_media_MediaPlayer_native_setup(JNIEnv *env, jobject thiz, jobject weak_t
     } else {
       ALOGE("QCMediaPlayer mediaplayer NOT present");
     }
+#endif
 
     // Stow our new C++ MediaPlayer in an opaque field in the Java object.
     setMediaPlayer(env, thiz, mp);
